File: app/src/test/java/openfoodfacts/github/scrachx/openfood/models/entities/allergen/AllergenResponseTest.java
package openfoodfacts.github.scrachx.openfood.models.entities.allergen;

import org.junit.Before;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static com.google.common.truth.Truth.assertThat;
import static openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_ENGLISH;
import static openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_FRENCH;
import static openfoodfacts.github.scrachx.openfood.models.entities.allergen.AllergenResponseTestData.PEANUTS_EN;
import static openfoodfacts.github.scrachx.openfood.models.entities.allergen.AllergenResponseTestData.PEANUTS_FR;
import static openfoodfacts.github.scrachx.openfood.models.entities.allergen.AllergenResponseTestData.UNIQUE_ALLERGEN_ID_1;

/**
 * Tests for {@link AllergenResponse}
 */
public class AllergenResponseTest {

    private Map<String, String> nameMap;

    @Before
    public void setUp(){
        nameMap = new HashMap<>();
    }

    @Test
    public void map_returnsMappedAllergen() {
        nameMap.put(LANGUAGE_CODE_ENGLISH, PEANUTS_EN);
        nameMap.put(LANGUAGE_CODE_FRENCH, PEANUTS_FR);
        AllergenResponse allergenResponse = new AllergenResponse(UNIQUE_ALLERGEN_ID_1, nameMap);
        Allergen allergen = allergenResponse.map();

        // TODO: expected this to be the allergen ID but a string to constructor actually makes it
        // the tag. Should update AllergenResponse to fix that.
        assertThat(allergen.getTag()).isEqualTo(UNIQUE_ALLERGEN_ID_1);
        assertThat(allergen.getNames()).hasSize(2);
        assertThat(allergen.getNames().get(0).getAllergenTag()).isEqualTo(UNIQUE_ALLERGEN_ID_1);
        assertThat(allergen.getNames().get(0).getName()).isEqualTo(PEANUTS_EN);
        assertThat(allergen.getNames().get(0).getLanguageCode()).isEqualTo(LANGUAGE_CODE_ENGLISH);
        assertThat(allergen.getNames().get(1).getAllergenTag()).isEqualTo(UNIQUE_ALLERGEN_ID_1);
        assertThat(allergen.getNames().get(1).getName()).isEqualTo(PEANUTS_FR);
        assertThat(allergen.getNames().get(1).getLanguageCode()).isEqualTo(LANGUAGE_CODE_FRENCH);
    }

    @Test
    public void map_returnsMappedAllergenWithWikiDataCode() {
        String wikiDataCode = "Q12345";
        AllergenResponse allergenResponse = new AllergenResponse(UNIQUE_ALLERGEN_ID_1, nameMap, wikiDataCode);
        Allergen allergen = allergenResponse.map();

        assertThat(allergen.getWikiDataId()).isEqualTo(wikiDataCode);
    }
}

File: app/src/test/java/openfoodfacts/github/scrachx/openfood/models/entities/allergen/AllergenResponseTest.kt
package openfoodfacts.github.scrachx.openfood.models.entities.allergen

import com.google.common.truth.Truth.assertThat
import openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_ENGLISH
import openfoodfacts.github.scrachx.openfood.models.LanguageCodeTestData.LANGUAGE_CODE_FRENCH
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

/**
 * Tests for [AllergenResponse]
 */
class AllergenResponseTest {
    lateinit var nameMap: MutableMap<String, String>
    @BeforeEach
    fun setUp() {
        nameMap = mutableMapOf()
    }

    @Test
    fun map_returnsMappedAllergen() {
        nameMap[LANGUAGE_CODE_ENGLISH] = PEANUTS_EN
        nameMap[LANGUAGE_CODE_FRENCH] = PEANUTS_FR
        val allergenResponse = AllergenResponse(UNIQUE_ALLERGEN_ID_1, nameMap)
        val allergen = allergenResponse.map()

        // TODO: expected this to be the allergen ID but a string to constructor actually makes it
        // the tag. Should update AllergenResponse to fix that.
        assertThat(allergen.tag).isEqualTo(UNIQUE_ALLERGEN_ID_1)
        assertThat(allergen.names).hasSize(2)
        assertThat(allergen.names[0].allergenTag).isEqualTo(UNIQUE_ALLERGEN_ID_1)
        assertThat(allergen.names[0].name).isEqualTo(PEANUTS_EN)
        assertThat(allergen.names[0].languageCode).isEqualTo(LANGUAGE_CODE_ENGLISH)
        assertThat(allergen.names[1].allergenTag).isEqualTo(UNIQUE_ALLERGEN_ID_1)
        assertThat(allergen.names[1].name).isEqualTo(PEANUTS_FR)
        assertThat(allergen.names[1].languageCode).isEqualTo(LANGUAGE_CODE_FRENCH)
    }

    @Test
    fun map_returnsMappedAllergenWithWikiDataCode() {
        val wikiDataCode = "Q12345"
        val allergenResponse = AllergenResponse(UNIQUE_ALLERGEN_ID_1, nameMap, wikiDataCode)
        val allergen = allergenResponse.map()
        assertThat(allergen.wikiDataId).isEqualTo(wikiDataCode)
    }
}